include/libopencm3/usb/audio.h

```diff
@@ -88,6 +90,143 @@ struct usb_audio_header_descriptor_body {
+	uint16_t wTerminalType;
+	uint8_t bAssocTerminal;
+	uint8_t bSourceID;
+	uint8_t iTerminal;
+} __attribute__((packed));
+
+/* Table 4-7: Feature Unit Descriptor (head) */
+struct usb_audio_feature_unit_descriptor_head {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bUnitID;
+	uint8_t bSourceID;
+	uint8_t bControlSize;
+	uint16_t bmaControlMaster; /* device can assume 16-bit, given highest
+				    * defined bit in spec is bit #9.
+				    * (it is thus required bControlSize=2) */
+	/* ... */
+} __attribute__((packed));
+
+/* Table 4-7: Feature Unit Descriptor (body) */
+struct usb_audio_feature_unit_descriptor_body {
+	/* ... */
+	uint16_t bmaControl;
+	/* ... */
+} __attribute__((packed));
+
+/* Table 4-7: Feature Unit Descriptor (tail) */
+struct usb_audio_feature_unit_descriptor_tail {
+	/* ... */
+	uint8_t iFeature;
+} __attribute__((packed));
+
+/* Table 4-7: Feature Unit Descriptor (2-channel)
+ *
+ * This structure is a convenience covering the (common) case where
+ * there are 2 channels associated with the feature unit
+ */
+struct usb_audio_feature_unit_descriptor_2ch {
+	struct usb_audio_feature_unit_descriptor_head head;
+	struct usb_audio_feature_unit_descriptor_body channel_control[2];
+	struct usb_audio_feature_unit_descriptor_tail tail;
+} __attribute__((packed));
+
+/* Table 4-19: Class-Specific AS Interface Descriptor */
+struct usb_audio_stream_interface_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bTerminalLink;
+	uint8_t bDelay;
+	uint16_t wFormatTag;
+} __attribute__((packed));
+
+/* Table 4-20: Standard AS Isochronous Audio Data Endpoint Descriptor */
+struct usb_audio_stream_endpoint_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bEndpointAddress;
+	uint8_t bmAttributes;
+	uint16_t wMaxPacketSize;
+	uint8_t bInterval;
+	uint8_t bRefresh;
+	uint8_t bSynchAddress;
+} __attribute__((packed));
+
+/* Table 4-21: Class-Specific AS Isochronous Audio Data Endpoint Descriptor */
+struct usb_audio_stream_audio_endpoint_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bmAttributes;
+	uint8_t bLockDelayUnits;
+	uint16_t wLockDelay;
+} __attribute__((packed));
+
+/*
+ * Definitions from the USB_AUDIO_FORMAT_ or usb_audio_format_ namespace come from:
+ * "Universal Serial Bus Device Class Definition for Audio Data Formats, Revision 1.0"
+ */
+
+/* Table 2-1: Type I Format Type Descriptor (head) */
+struct usb_audio_format_type1_descriptor_head {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bFormatType;
+	uint8_t bNrChannels;
+	uint8_t bSubFrameSize;
+	uint8_t bBitResolution;
+	uint8_t bSamFreqType;
+	/* ... */
+} __attribute__((packed));
+
+/* Table 2-2: Continuous Sampling Frequency */
+struct usb_audio_format_continuous_sampling_frequency {
+	/* ... */
+	uint32_t tLowerSamFreq : 24;
+	uint32_t tUpperSamFreq : 24;
```

Comment:
is this really the best way?  bitfields have notorious portability problems, so I tend to just avoid them outright.  I can't relaly think of anythign better though, and it _works_ so... probably the best way?

Reply:
I mean, we could use an array of 3 uint8_t's with a special macro; but I would prefer the bitfield

Reply:
yeah, it's good, let's keep it as is.